### DISCUSSION TITLE
expo-av  how to play audio with URL

### DISCUSSION THREAD

**Discussion by hayata-suenaga** [expo/expo#18869] [Archived: Questions]
I'm passing a url to an audio file on the internet to `createAsync`, but the audio is not playing. When I use an audio file that lives alongside the source code with `require()`, the audio plays without problem.

```
  const { sound } = await Audio.Sound.createAsync({uri: url});
  await sound.playAsync();
  await sound.unloadAsync();
```

Is there any specific steps needed for playing audio on the internet?

**Comment by barthap** [comment]
Looking at your code, you're trying to play it right after creation, so the sound object isn't loaded from URL yet (network stuff can take time).
The thing you should do is to [listen for `onPlaybackStatusUpdate` events](https://docs.expo.dev/versions/latest/sdk/audio/#setonplaybackstatusupdateonplaybackstatusupdate) and check if [`isLoaded == true`](https://docs.expo.dev/versions/latest/sdk/av/#avplaybackstatussuccess). You may also be interested in the `isBuffering` prop.
```js
let canPlay = false; // this should be stored as some kind of state

const { sound } = await Audio.Sound.createAsync({uri: url});
sound.setOnPlaybackStatusUpdate(status => {
  if (status.isLoaded) canPlay = true;
});

// ...

if (canPlay) await sound.playAsync();
```

The advanced player example [can be found here](https://github.com/expo/expo/blob/b79b3bc03d494a542ccff3447fba0d71a6965d00/apps/native-component-list/src/screens/AV/AudioScreen.tsx#L27).

**Reply by finom** [reply]
Doesn't work for me :(

**Comment by ser-emejia** [comment]
> I'm passing a url to an audio file on the internet to `createAsync`, but the audio is not playing. When I use an audio file that lives alongside the source code with `require()`, the audio plays without problem.
> 
> ```
>   const { sound } = await Audio.Sound.createAsync({uri: url});
>   await sound.playAsync();
>   await sound.unloadAsync();
> ```
> 
> Is there any specific steps needed for playing audio on the internet?

Did you find  a solution?

**Reply by mprinceb** [reply]
```js
async function playSound() {
    if (url) {
      const { sound } = await Audio.Sound.createAsync(
        {
          uri: url,
        },
        { shouldPlay: true },
      );
      await sound.playAsync();
    }
  }
```
**this worked fine for me**

**Comment by fave-xuerong** [comment]
I also encountered the same problem. How did you solve it at last

**Comment by yoginsurya** [comment]
For me, I was able to get it to work on IOS by setting the shouldPlay option and playsInSilentModeIOS to true. See below – the code was part of an AudioPlayer component and the playSound was triggered by a button press:

```typescript
const [isPlaying, setIsPlaying] = useState(false);
const [sound, setSound] = useState<Audio.Sound | null>(null);
const url =
    "example.com/audio";

async function playSound() {
    console.log("Loading Sound");
    await Audio.setAudioModeAsync({
        playsInSilentModeIOS: true,
    });
    const { sound } = await Audio.Sound.createAsync(
        { uri: url },
        { shouldPlay: true }
    );
    setSound(sound);

    console.log("Playing Sound");
    await sound.playAsync();
}

useEffect(() => {
    return sound
        ? () => {
              console.log("Unloading Sound");
              setIsPlaying(false);
              sound.unloadAsync();
              setSound(null);
          }
        : undefined;
}, [sound]);
```

**Comment by webkokri** [comment]
const MusicPlayer = () => {

    const [sound, setSound] = useState(null);
  const [isPlaying, setIsPlaying] = useState(false);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    Audio.setAudioModeAsync({
      staysActiveInBackground: true,
      playsInSilentModeIOS: true,
      // interruptionModeIOS: InterruptionModeIOS.DuckOthers,
      // interruptionModeAndroid: InterruptionModeAndroid.DuckOthers,
      shouldDuckAndroid: true,
      playThroughEarpieceAndroid: true,
  });
    const loadSound = async () => {
      setLoading(true);
      const { sound } = await Audio.Sound.createAsync(
        { uri: 'https://url/file.mp3' },
        { shouldPlay: false }
      );
      setSound(sound);
      setLoading(false);
      // Set the volume of the sound to a high level (1.0 = 100%)
      await sound.setVolumeAsync(1.0);
    };

    loadSound();
  }, []);

  const play = async () => {
    if (sound!== null) {
      await sound.playAsync();
      setIsPlaying(true);
    }
  };

  const pause = async () => {
    if (sound!== null) {
      await sound.pauseAsync();
     // await sound.stopAsync();
      setIsPlaying(false);
    }
  };
